### PR TITLE
implement to extract success-traj

### PR DESCRIPTION
**Pull request by ByungJi-Park**
1. I added new attribute(self._num_sim_success) in class POMCPOW to collect success trajectories in simulation.
2. Then, whenever trajectories in simulation succeed to reach a goal, they are appended.
3. Furthermore, I made new file(data_process.py) for data processing.
data type for action -> (type:string, target:string, pos:tuple, orn:tuple)
data type for observation -> (img: list, mask:dict[obj:str, list[list[]]])
data type for reward -> float
4. I tried to finish implementing dataloader, if you are getting better I want to talk about this.

### CODE REVIEW

File: Simulation/pybullet_env/fetching_POMDP_primitive_object.py [hunk 1]
```diff
@@ -901,43 +853,33 @@ def update(self, real_action: FetchingAction, real_observation: FetchingObservat
...
                 "robot": p.robot,
                 "object": p.object,
                 "weight": v
                 })
-        
+            '''
+
+            for p, v in zip(particles, weights):
+                next_particles[p] = v
+
```
Comment (ajy8456) [inline Simulation/pybullet_env/fetching_POMDP_primitive_object.py:882]: Also, why did you change this part?

File: Simulation/pybullet_env/fetching_POMDP_primitive_object.py [hunk 2]
```diff
@@ -994,26 +936,26 @@ def make_belief(bc, env, robot, config, num_particle):
...
+    #     init_belief[p] = v
 
-    # for p in particles:
-    #     init_belief[p] = 1.0
+    for p in particles:
+        init_belief[p] = 1.0
 
-    return WeightedParticles(init_belief)
\ No newline at end of file
+    return WeightedParticles(init_belief)
```
Comment (ajy8456) [inline Simulation/pybullet_env/fetching_POMDP_primitive_object.py]: Too

File: Simulation/pybullet_env/fetching_problem_primitive_object.py [hunk 3]
```diff
@@ -207,14 +211,21 @@ def main(config, n):
         
         if (termination == "success") or (termination == "fail"):
             break
-    
+    # Store success trajectory in simulation
+    with open('data/1.18/prev_success_traj.pickle', 'wb') as f:
+        pickle.dump(planner._success_traj_sim, f)    
+    success_trajs_simulation = process_success_traj_data(planner._success_traj_sim, agent.env.uid_to_target)
+    with open('data/1.18/post_success_traj.pickle', 'wb') as f:
+        pickle.dump(success_trajs_simulation, f)
```
Comment (ajy8456) [inline Simulation/pybullet_env/fetching_problem_primitive_object.py:219]: Why did you save both data before and after processing?

Comment (ajy8456) [reply Simulation/pybullet_env/fetching_problem_primitive_object.py:219]: It seems better to make it possible to set whether to save the data or not according to the configuration.

File: Simulation/pybullet_env/POMCPOW.py [hunk 4]
```diff
@@ -179,7 +182,8 @@ def _simulate(self, state: State, history: Tuple, root: VNode, depth: int): # ro
                     print('_simulate() end')
                     if termination == "success":
                         self._num_sim_success += 1
-            
+                        self._success_traj_sim.append(history)
```
Comment (ajy8456) [inline Simulation/pybullet_env/POMCPOW.py:185]: It seems better to make it possible to set whether to save the data or not according to the configuration.

File: Simulation/pybullet_env/data_process.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+def seg_uid_to_target(x, uid_to_target) -> dict:
+        for uid in list(x.keys()):
+            # Delete non-objects
+            if uid not in uid_to_target:
+                x.pop(uid)
+                continue
+            # Transform into targets
+            obj = uid_to_target[uid]
+            x[obj] = x.pop(uid)
+        return x
```
Comment (ajy8456) [inline Simulation/pybullet_env/data_process.py:20]: This function will only work for the simplified version of the problem. I think we should consider this when refactoring.

File: Simulation/pybullet_env/data_process.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    for success_traj in history:
+        data_action = []
+        data_observation = []
+        data_reward = []
+        for a, o, *rest in success_traj:
+            # if history has next state
+            if len(rest) == 2:
+                r = rest[1]
+            else:
+                r = rest[0]
```
Comment (ajy8456) [inline Simulation/pybullet_env/data_process.py:34]: Is there any case that history has next state?

File: Simulation/pybullet_env/data_process.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            data_observation.append((o.observation.tolist(), seg_uid_to_target(o.seg_mask, uid_to_target)))
+            data_reward.append(r)
+        # Sanity Check
+        if (len(data_action) != len(data_observation)) or (len(data_action) != len(data_reward)) or (len(data_reward) != len(data_observation)):
+            raise ValueError("Length mismatched!")
+        data = {"action": data_action, "observation": data_observation, "reward": data_reward}
+        history_list.append(data)
+    return history_list
+
+class fetching_data(torch.utils.data.Dataset):
```
Comment (ajy8456) [inline Simulation/pybullet_env/data_process.py:45]: Class names in our project should be pascal case.
ex. fetching_data: snake case, FetchingData: pascal case, fetchingData: camel case

File: Simulation/pybullet_env/data_process.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    
+    def __len__(self):
+        return len(self.data)
+
+def set_dataloader(data_path, train, batch_size):
+    trainset = fetching_data(data_path, train)
+    trainloader = torch.utils.data.DataLoader(trainset, batch_size=batch_size,
+                                              shuffle=True, num_workers=2)
+    return trainloader
+
```
Comment (ajy8456) [inline Simulation/pybullet_env/data_process.py:68]: Did you test this data loader? I think it will be fail to make batch with different length of data.

**Review by ajy8456 — CHANGES_REQUESTED** [review]
I think you should mainly check the items below:

- Creating an option to save data or not
- Unnecessary modifications to `fetching_POMDP_primitive_object.py`
- Implementing a code to verify that the data loader is working

File: Simulation/pybullet_env/POMCPOW.py [hunk 9]
```diff
@@ -49,6 +49,10 @@ def __init__(self, pomdp: POMDP, config):
         self._evaluate_node = config["plan_params"]["evaluate_node"]
         self._simulator = config["plan_params"]["simulator"]
         
+        # Edit
+        # Store success trajectory in simulation
+        self._success_traj_sim = []
```
[the remark this answers is not in the document]
Comment (ByungJi-Park) [reply Simulation/pybullet_env/POMCPOW.py:54]: Oh, I'm sorry that I typed wrong.

File: Simulation/pybullet_env/fetching_problem_primitive_object.py (hunk 3 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/fetching_problem_primitive_object.py:219]: This is for debugging, but I forgot to remove those codes.
Furthermore, let me implement to set storage option in configuration.

File: Simulation/pybullet_env/POMCPOW.py (hunk 4 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/POMCPOW.py:185]: You mean saving success trajectory history?
If it's right, let me implement.

File: Simulation/pybullet_env/data_process.py (hunk 5 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/data_process.py:20]: Okay, let me consider about it.

File: Simulation/pybullet_env/data_process.py (hunk 6 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/data_process.py:34]: **history += ((action, observation, next_state, reward),)**

            if (termination == "success") or (termination == "fail"):

                total_reward = reward
        
                root.num_visits += 1
                root[action].num_visits += 1
                root[action].value = root[action].value + (total_reward - root[action].value) / (root[action].num_visits)
                    
                # ======================== Debug =========================
                if self.config['project_params']['debug']['get_data']:
                    print('_simulate() end')
                    if termination == "success":
                        self._num_sim_success += 1
                        **self._success_traj_sim.append(history)**

When referring to the above code, history has 4 factors which are (action, observation, next_state, reward).
That's why history might have next state, so I wanted to ask there is another way to implement...

File: Simulation/pybullet_env/data_process.py (hunk 7 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/data_process.py:45]: Oh, I didn't know that. Let me fix it.

File: Simulation/pybullet_env/data_process.py (hunk 8 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/data_process.py:68]: It will not fail, because data type is list which each success trajectory's type is dictionary. -> [dict, dict, dict ... dict]  
Therefore, even though the length of each success trajectory is different, it's fine. (Also, I have already tested.)
I'd like to ask if I did it right.

File: Simulation/pybullet_env/fetching_POMDP_primitive_object.py [hunk 10]
```diff
@@ -274,114 +273,67 @@ def get_sensor_observation(self, next_state: FetchingState, action: FetchingActi
...
+        # # ======================== Debug =========================
+        # if self.config['project_params']['debug']['get_data']:
+        #     print('ObservationModel.probability() end')
+        #     check_time = time.time() - check_time
+        #     debug_data['ObservationModel.probability()'].append(check_time)
+        
+        if log:
+            return log_prob
+        else:
+            return prob
```
[the remark this answers is not in the document]
Comment (ByungJi-Park) [reply Simulation/pybullet_env/fetching_POMDP_primitive_object.py:336]: Actually, I never touched those parts, so it's very weird situation.
One possibility is git clone wrong part at first.
Thus, can I just modify while referring to your code in main branch?

File: Simulation/pybullet_env/fetching_POMDP_primitive_object.py (hunk 1 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/fetching_POMDP_primitive_object.py:882]: This is also same situation.

File: Simulation/pybullet_env/fetching_POMDP_primitive_object.py (hunk 2 above)
Comment (ByungJi-Park) [reply Simulation/pybullet_env/fetching_POMDP_primitive_object.py]: This is also same situation.

File: Simulation/pybullet_env/fetching_POMDP_primitive_object.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ajy8456) [reply Simulation/pybullet_env/fetching_POMDP_primitive_object.py:336]: You should commit only parts that you modified or implemented.

File: Simulation/pybullet_env/POMCPOW.py (hunk 4 above)
Comment (ajy8456) [reply Simulation/pybullet_env/POMCPOW.py:185]: It is the same as the comment above.

File: Simulation/pybullet_env/data_process.py (hunk 6 above)
Comment (ajy8456) [reply Simulation/pybullet_env/data_process.py:34]: You'd better leave an comment to make it easier for others to understand.

File: Simulation/pybullet_env/data_process.py (hunk 7 above)
Comment (ajy8456) [reply Simulation/pybullet_env/data_process.py:45]: I should have told you. It would be helpful if you write code considering uniformity when doing programming collaboration.

File: Simulation/pybullet_env/data_process.py (hunk 8 above)
Comment (ajy8456) [reply Simulation/pybullet_env/data_process.py:68]: What is the structure of the data processed through the data loader? As far as I know, Pytorch dataloader basically puts data in the list unless you implement processing to stack data in batches of different lengths. One of the roles of the Pytorch dataloader is to match the interface of the input that enters the network. We have to stack the data in batches to learn. So the data loader has to make the sampled data the same size, stack them in batches, and make them into one tensor.
